of-builder/main.go

```diff
@@ -88,17 +92,50 @@ func build(w http.ResponseWriter, r *http.Request) ([]byte, error) {
 	if cfg.Frontend == "" {
 		cfg.Frontend = "tonistiigi/dockerfile:v0"
 	}
-
-	cmd := exec.Command("buildctl", "--debug", "build", "--frontend=gateway.v0", "--frontend-opt=source="+cfg.Frontend, "--local=context="+filepath.Join(tmpdir, "context"), "--local=dockerfile="+filepath.Join(tmpdir, "context"), "--no-progress", "--exporter=image",
-		"--exporter-opt=name="+cfg.Ref, "--exporter-opt=push=true", "--exporter-opt=registry.insecure=true")
-	env := os.Environ()
-	env = append(env, "BUILDKIT_HOST=tcp://of-buildkit:1234")
-	cmd.Env = env
-	cmd.Stdout = os.Stdout
-	cmd.Stderr = os.Stderr
-	if err := cmd.Run(); err != nil {
+	contextDir := filepath.Join(tmpdir, "context")
+	solveOpt := client.SolveOpt{
+		Exporter: "image",
+		ExporterAttrs: map[string]string{
+			"name":              cfg.Ref,
+			"push":              "true",
+			"registry.insecure": "true",
+		},
+		LocalDirs: map[string]string{
+			"context":    contextDir,
+			"dockerfile": contextDir,
+		},
+		Frontend: "gateway.v0",
+		FrontendAttrs: map[string]string{
+			"source": cfg.Frontend,
+		},
+		// ~/.docker/config.json could be provided as Kube or Swarm's secret
+		Session: []session.Attachable{authprovider.NewDockerAuthProvider()},
+	}
+	c, err := client.New("tcp://of-buildkit:1234", client.WithBlock())
+	if err != nil {
+		return nil, err
+	}
+	ch := make(chan *client.SolveStatus)
+	eg, ctx := errgroup.WithContext(context.Background())
+	eg.Go(func() error {
+		return c.Solve(ctx, nil, solveOpt, ch)
+	})
+	eg.Go(func() error {
+		for s := range ch {
+			for _, v := range s.Vertexes {
+				logrus.Debugf("vertex: %s %s %v %v", v.Digest, v.Name, v.Started, v.Completed)
```

Comment:
@AkihiroSuda I've used Go's `log` in the other packages due to headaches with logrus and vendoring - after their upper/lowercase rename it made things difficult. Is logrus required?

Reply:
updated to use `log`, but `logrus` is still vendored because other vendored packages import logrus